src/datasets/mnist.nim

```diff
@@ -95,19 +97,97 @@ proc read_mnist_labels*(labelsPath: string): Tensor[uint8] {.noInit.}=
+
+  let stream = newGzFileStream(imgsPath, mode = fmRead)
+  return read_mnist_images(stream)
 
-  let stream = newFileStream(labelsPath, mode = fmRead)
+proc read_mnist_labels*(stream: Stream): Tensor[uint8] {.noInit.}=
+  ## Load MNIST labels into a Tensor[uint8] from a file
+  ## Input:
+  ##   - A stream of MNIST labels data
+  ##
+  ## Returns:
+  ##   - A tensor of labels with shape (N)
+  ##     - N, number of images
   defer: stream.close
 
   let magic_number = stream.readInt32BE
-  doAssert magic_number == 2049'i32, "This file is not a MNIST labels file, did you forget to decompress it?"
+  doAssert magic_number == 2049'i32, "This file is not a MNIST labels file."
 
   let
     n_labels = stream.readInt32BE.int
 
   result = newTensorUninit[uint8](n_labels)
   discard stream.readData(result.get_data_ptr, result.size)
+
+proc read_mnist_labels*(labelsPath: string): Tensor[uint8] {.noInit.}=
+  ## Load MNIST labels into a Tensor[uint8] from a file
+  ## Input:
+  ##   - A path to a MNIST labels file
+  ##
+  ## Returns:
+  ##   - A tensor of labels with shape (N)
+  ##     - N, number of images
+  ## MNIST data can be downloaded here: http://yann.lecun.com/exdb/mnist/
+  if not existsFile(labelsPath):
+    raise newException(IOError, "MNIST labels file \"" & labelsPath & "\" does not exist")
+
+  let stream = newGzFileStream(labelsPath, mode = fmRead)
+  return read_mnist_labels(stream)
+
+proc get_mnist_files(): seq[string] =
+  ## Produces the file names for the MNIST dataset
+  ## Returns:
+  ## - seq[string] of filenames
+  return @[
+    "train-images-idx3-ubyte.gz",
+    "train-labels-idx1-ubyte.gz",
+    "t10k-images-idx3-ubyte.gz",
+    "t10k-labels-idx1-ubyte.gz",
+  ]
+
+proc download_mnist_files(files: seq[string]) =
+  ## Download the MNIST files from http://yann.lecun.com/exdb/mnist/
+  ## It will download the files to the current directory.
+  let
+    mnist_domain = "http://yann.lecun.com/exdb/mnist/"
+    client = newHttpClient()
+
+  for f in files:
+    let url = fmt"{mnist_domain}{f}"
+    client.downloadFile(url, tmp & f)
+
+proc delete_mnist_files(files: seq[string]) =
+  ## Deletes the downloaded MNIST files.
+  for f in files:
+    discard tryRemoveFile(tmp & f)
+
+proc load_mnist*(): mnist =
+  ## Loads the MNIST dataset into a tuple with fields:
+  ## - train_images
+  ## - train_labels
+  ## - test_images
+  ## - test_labels
+  ##
+  ## This proc will:
+  ## - download the files
+  ## - unzip them
+  ## - load into a tuple
+  ## - delete the downloaded files
+  let
+    mnist_files = get_mnist_files()
+    tmp_files = map(mnist_files, (s: string) -> (string) => (tmp & s))
+
+  download_mnist_files(mnist_files)
+
+  # Training
+  result.train_images = read_mnist_images(tmp_files[0])
+  result.train_labels = read_mnist_labels(tmp_files[1])
+
+  # Testing
+  result.test_images = read_mnist_images(tmp_files[2])
+  result.test_labels = read_mnist_labels(tmp_files[3])
+
+  delete_mnist_files(tmp_files)
```

Comment:
MNIST is small but I guess for heavy datasets we should provide a way to save permanently and to manage the cached datasets and ML models

Reply:
Yeah, that would be cool